src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/InMemoryPartitionedAppTokenCacheAccessor.cs

```diff
@@ -0,0 +1,200 @@
+using Microsoft.Identity.Client.Core;
+
+namespace Microsoft.Identity.Client.PlatformsCommon.Shared
+{
+    /// <summary>
+    /// Stores tokens for an application.
+    /// Partitions the access token collection by a key of client ID with tenant ID.
+    /// App metadata collection is not partitioned.
+    /// Refresh token, ID token, and account related methods are no-op.
+    /// </summary>
+    internal class InMemoryPartitionedAppTokenCacheAccessor : ITokenCacheAccessor
+    {
+        // perf: do not use ConcurrentDictionary.Values as it takes a lock
+        internal /* internal for test only */ readonly ConcurrentDictionary<string, ConcurrentDictionary<string, MsalAccessTokenCacheItem>> AccessTokenCacheDictionary =
+            new ConcurrentDictionary<string, ConcurrentDictionary<string, MsalAccessTokenCacheItem>>(1, 1);
+
+        internal /* internal for test only */  readonly ConcurrentDictionary<string, MsalAppMetadataCacheItem> AppMetadataDictionary =
+           new ConcurrentDictionary<string, MsalAppMetadataCacheItem>(1, 1);
+
+        protected readonly ICoreLogger _logger;
+
+        public InMemoryPartitionedAppTokenCacheAccessor(ICoreLogger logger)
+        {
+            _logger = logger ?? throw new System.ArgumentNullException(nameof(logger));
+        }
+
+        #region Add
+        public void SaveAccessToken(MsalAccessTokenCacheItem item)
+        {
+            var partitionKey = CacheKeyFactory.GetClientCredentialKey(item.ClientId, item.TenantId);
+            string itemKey = item.GetKey().ToString();
+            // if a conflict occurs, pick the latest value
+            AccessTokenCacheDictionary
+                .GetOrAdd(partitionKey, new ConcurrentDictionary<string, MsalAccessTokenCacheItem>())[itemKey] = item;
+        }
+
+        /// <summary>
+        /// This method is not supported for the app token cache because
+        /// there are no refresh tokens, ID tokens, or accounts in a client credential flow
+        /// </summary>
+        public void SaveRefreshToken(MsalRefreshTokenCacheItem item)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <summary>
+        /// This method is not supported for the app token cache because
+        /// there are no refresh tokens, ID tokens, or accounts in a client credential flow
+        /// </summary>
+        public void SaveIdToken(MsalIdTokenCacheItem item)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <summary>
+        /// This method is not supported for the app token cache because
+        /// there are no refresh tokens, ID tokens, or accounts in a client credential flow
+        /// </summary>
+        public void SaveAccount(MsalAccountCacheItem item)
+        {
+            throw new NotSupportedException();
+        }
+
+        public void SaveAppMetadata(MsalAppMetadataCacheItem item)
+        {
+            string key = item.GetKey().ToString();
```

Comment:
`item.GetKey().` null check?

Reply:
can't be null. An assert should be enough, but a check if too much.